synapse/app/generic_worker.py

```diff
@@ -377,22 +378,37 @@ def _listen_http(self, listener_config: ListenerConfig) -> None:
 
         root_resource = create_resource_tree(resources, OptionsResource())
 
-        _base.listen_tcp(
-            bind_addresses,
-            port,
-            SynapseSite(
-                "synapse.access.http.%s" % (site_tag,),
-                site_tag,
-                listener_config,
-                root_resource,
-                self.version_string,
-                max_request_body_size=max_request_body_size(self.config),
-                reactor=self.get_reactor(),
-            ),
+        site = SynapseSite(
+            "synapse.access.%s.%s" % ("https" if tls else "http", site_tag),
+            site_tag,
+            listener_config,
+            root_resource,
+            self.version_string,
+            max_request_body_size=max_request_body_size(self.config),
             reactor=self.get_reactor(),
         )
-
-        logger.info("Synapse worker now listening on port %d", port)
+        # This has same logic as synapse/app/homeserver.py where we choose between
+        # listen_ssl and listen_tcp based on the tls configuration flag in listener
+        # config.
```

Comment:
Can we abstract this out so that we can call something like:

```python
self._listen_something(listener_config, root_resource)
```

This can then do the decision of TLS or not and possibly create the `site` (alternately, pass the `site` instead of `root_resource` in).

(Name to be determined! 😄 )